### ISSUE TITLE
Performance issue with the redundancy load

### ISSUE THREAD

**Issue by yyzdtccjdtc** [utkarshkukreti/diff.rs#15]
https://github.com/utkarshkukreti/diff.rs/blob/5de4cb51da9349ff5af96766af03d54329baa7f0/src/lib.rs#L65-L81

For line 71 here, although the right_skip has been defined in line 68 with the clone(), skip(), and take() functions, these three functions will be actually called in line 71 because of the lazy evaluation. Because the same functions keep being called on the same string inside the for loop, this causes a lot of redundant memory load operations. Based on my test with the rustfmt applications (https://github.com/rust-lang/rustfmt), this can be avoided if we change the right_skip to a vector and pass it by reference to the inside of the for loop. After this optimization, with my test input file, there is a 7.12x speedup and the execution time is reduced from 12.56 seconds to 1.54 seconds.

Here's the modified code

```
        let table: Vec<Vec<u32>> = {
        let mut table = vec![vec![0; right_diff_size + 1]; left_diff_size + 1];
        let mut left_skip = left.clone().skip(leading_equals).take(left_diff_size).enumerate();
        let mut right_skip = right.clone().skip(leading_equals).take(right_diff_size).enumerate().collect::<Vec<_>>();

        for (i, l) in left_skip {
            for (j, r) in &right_skip {
                table[i + 1][*j + 1] = if l == *r {
                    table[i][*j] + 1
                } else {
                    std::cmp::max(table[i][*j + 1], table[i + 1][*j])
                };
            }
        }

        table
    };
```

**Comment by utkarshkukreti** [comment]
Hi @yyzdtccjdtc.

That's a good find. I believe the compiler is able to optimize the code for slice iterators; your test case is probably diffing strings or other custom iterator. I pushed an alternative solution (https://github.com/utkarshkukreti/diff.rs/commit/0f0aa580f197af7dab9ab905568f70a2387e23bc) which does not allocate any `Vec<T>` but significantly speeds up a sample test case I came up with (also in the same commit). Can you try running your test input with this change and see how it performs?

**Comment by yyzdtccjdtc** [comment]
> Hi @yyzdtccjdtc.
> 
> That's a good find. I believe the compiler is able to optimize the code for slice iterators; your test case is probably diffing strings or other custom iterator. I pushed an alternative solution ([0f0aa58](https://github.com/utkarshkukreti/diff.rs/commit/0f0aa580f197af7dab9ab905568f70a2387e23bc)) which does not allocate any `Vec<T>` but significantly speeds up a sample test case I came up with (also in the same commit). Can you try running your test input with this change and see how it performs?

Hi @utkarshkukreti 
Sorry for the late reply. Yes you are right, I have found this performance issue because your crate were used in the rust's official format check app to diffing code (strings like you said). And I have tested your new code and it gives about a 4% speedup. I see that you replaced skip() function to advance_by(), but I believe there are still some redundancy operation with take() and clone() functions, so maybe allocate an immutable vector and reference would be better. Can you try it to see if it works?

**Comment by utkarshkukreti** [comment]
I profiled some test cases after our last messages and came up with some more changes in #16. Could you try those out and see if they perform better on your data?

**Comment by yyzdtccjdtc** [comment]
@utkarshkukreti 
Yes, your new code works as fast as my approach now! I believe it's because you replaced the skip() function with the index with right_skip. Good work!

**Comment by utkarshkukreti** [comment]
This has been released in 0.1.13. Thanks for suggesting the improvement!
